synapse/handlers/event_auth.py

```diff
@@ -326,13 +328,13 @@ async def get_rooms_that_allow_join(
 
         # If allowed is of the wrong form, then only allow invited users.
         allow_list = join_rules_event.content.get("allow", [])
-        if not isinstance(allow_list, list):
+        if not isinstance(allow_list, (list, tuple)):
             return ()
 
         # Pull out the other room IDs, invalid data gets filtered.
         result = []
         for allow in allow_list:
-            if not isinstance(allow, dict):
+            if not isinstance(allow, (immutabledict, dict)):
```

Comment:
To match the order of the `(list, tuple)` pattern:

```suggestion
            if not isinstance(allow, (dict, immutabledict)):
```